### ISSUE TITLE
Does Yarn have API limitations?

### ISSUE THREAD

**Issue by tsabat** [yarnpkg/yarn#5697]
## TLDR

Will we get throttled if we make regular calls to the yarn API?

## Intro

Hi yarn team!

I'm a founder of [CodePen](https://codepen.io) and I have a question about using the Yarn API.

We're hacking on yarn to allow CodePen users to build dll files from npm dependencies on-demand which can then be used within Pens/Projects.  We're using a bunch of the cool features that yarn has opened up, like the yarn offline mirror, as well as hacking the yarn source to get a yarn.lock without actually pulling the dependencies.  

In short, we're using Yarn in a pretty custom way to make CodePen user's lives easier.  As we approach production, there's one question that remains, regarding throttling against the yarn api, and whether we'll be rate-limited or throttled if we try to do what we describe below.

## Project Details

Here's our high-level description of the project steps, so the question below comes with some context.

1. user chooses the packages she wants
1. the request is sent to a build farm
1. the build farm pre-builds a yarn.lock to determine dependencies
1. an offline mirror is populated with any dependencies the build farm does not already have
1. the the offline mirror is mounted within a docker container
1. a `yarn install --offline` call inside the docker container
1. a `yarn build dist` call is made to build a dll file, which is uploaded to s3

## Question

Finally, here's the question.  There are two places where we'll be calling out to [yarnpkg.com](http://yarnpkg.com) and we want to be sure we're not going to get throttled or banned for repeatedly calling to your APIs.

In step 3 of the Project Details above, we use a custom build of `yarn install` to build a yarn.lock, but we bail out before we actually fetch the packages.

Then, in step 4 of the Project Details above, we're making http calls to yarn's api to build our own on-disk cache.  An example call would be to

    https://registry.yarnpkg.com/spdx-license-ids/-/spdx-license-ids-3.0.0.tgz

We worry that making repeated calls to [registry.yarnpkg.com](http://registry.yarnpkg.com) will get us rate-limited or banned.  In step 4, we will be building our own on-disk cache of these libraries, so while the volume of requests may be heavy at first, they should taper off as our disk cache fills up.  However, step 3 will happen every time a person asks to build a DLL that does not already exist.

We're unsure how popular this feature will be, but with millions of users, we can expect a few calls per second, with tapers during US evenings. Is there any chance we'll be rate-limited?  Should we send a user-agent to help identify ourselves, so if an abuse team looks at our traffic it won't assume we're bad actors?

In short, we're looking for Yarn's blessing to continue down this path, with the agreement that we'll do everything we can to reduce the load on your servers.

**Comment by arcanis** [comment]
I *think* it should be fine. The two things that might matter are:

- If I'm not mistaken we are behind cloudflare, so maybe they have some anti-DDoS protection you'll want to avoid triggering.

- The registry is currently forwarding all the requests to the npm registry (we don't cache data as far as I know), so the load will ultimately be on them.

I think @Daniel15 or @kittens have more details on how Cloudflare has been setup.

**Comment by tsabat** [comment]
That's good news.  The fact that you're behind cloudflare makes the proxied requests to NPM a non-issue.  But, knowing what our limitations are for cloudflare up-front would be helpful.  @Daniel15 or @kittens would you mind weighing in on this?  Worst case, if we're rate-limited, we could attempt to call npm instead of yarn and get rate-limited by them.  ;)

Any other suggestions, particularly regarding user-agents, or anything else that comes to mind?

**Comment by arcanis** [comment]
> Any other suggestions, particularly regarding user-agents, or anything else that comes to mind?

I guess a custom user agent would be nice in case we latter want to make a breakdown of the download count per source, but there's no requirement per se.

**Comment by Daniel15** [comment]
As far as I know, `registry.yarnpkg.com` is simply a reverse proxy to `registry.npmjs.com`, and I don't think Cloudflare does any rate limiting by default (although a *huge* number of hits from a small IP range may trigger their anti-DDoS mechanisms, but that's true of any Cloudflare site).

> we can expect a few calls per second, with tapers during US evenings. Is there any chance we'll be rate-limited?

To be honest, I don't think a few calls per second would even be noticeable to Cloudflare 😛 

It looks like right now we're getting around 2 billion requests per week, so I don't think an increase of this magnitude would hurt anything. 

Out of curiosity, what does "dll files" mean in this context:
> We're hacking on yarn to allow CodePen users to build dll files from npm dependencies

I assume it doesn't mean the 'regular' definition of dll files (ie. dynamic libraries on Windows)

**Comment by Daniel15** [comment]
I just checked our Cloudflare configuration and we don't have any of the security feature (rate limiting, etc.) enabled, so I don't think you'll hit any issues.

**Comment by tsabat** [comment]
Thanks for the details @Daniel15.  Sounds like CloudFlare won't be an issue.

DLL is a woefully bad naming convention for [WebPack DLLs](https://webpack.js.org/plugins/dll-plugin/).  Considering all the history of that name, I'm surprised they chose it.  To avoid that marketing problem, we're referring to them as Bundles in our documentation.
